Add TTL cache to redfish discover method

dashmage opened the pull request:
The `redfish.discover_ssdp()` method takes a very long time to query the redfish services. This causes the scrape timeout to be exceeded for Prometheus. In order to solve this, a TTL cache has been added. A logging message has also been added for debugging.

Additionally, the `redfish.redfish_client` call to create a redfish object has been provided with the `timeout` and `max_retry` arguments in order to not have the query take too long in case of incorrect redfish credentials.

Fixes #33, canonical/hardware-observer-operator#55

dashmage commented:
Not able to run `tox -e lint`. Raised another issue -- #35

rgildein reviewed (commented):
Can we provide simple unit tests to ensure that cache is working as we expected? Also there are no tests for `_get_sensor_data` function and it's probably out of scope of this PR, but it would be nice if you could add it also.

jneo8 commented on prometheus_hardware_exporter/collectors/redfish.py:
```diff
@@ -4,10 +4,26 @@
 logger = getLogger(__name__)
 
+# timeout in seconds for initial connection
+REDFISH_CLIENT_TIMEOUT = 3
```
**nitpick:** 

Instead of putting settings here, how about loading them from the config?
We can provide a default value also.

dashmage replied:
I think that it's best we keep these configs directly in `redfish.py` since it would be easier to refer them and change it here rather than going to a different file. 

The configs for `sasircu.py` collector are also inside the same file.

jneo8 replied:
I would like to change the settings with config instead of the source code change. It will be useful when debugging or turning.

jneo8 commented on prometheus_hardware_exporter/collectors/redfish.py:
```diff
@@ -44,21 +60,32 @@ def _get_sensor_data(self, host: str, username: str, password: str) -> Optional[
+    @cached(cache=TTLCache(maxsize=CACHE_MAXSIZE, ttl=CACHE_TTL))
     def discover(self) -> bool:
         """Return true if redfish is been discovered."""
+        logger.info("Discovering redfish services")
```
**nitpick (non-blocking):** 

Can we also put log to show the discover result?

dashmage replied:
Good idea. I'll also add another info log in case no redfish services have been discovered.

jneo8 reviewed (changes requested):
Just as @rgildein said. Please provide unit tests in the PR.

sudeephb reviewed (commented):
After adding unit tests, as already mentioned in other comments, everything else LGTM.

sudeephb commented on prometheus_hardware_exporter/__main__.py:
```diff
@@ -112,6 +118,24 @@ def parse_command_line() -> argparse.Namespace:
+    )
+    parser.add_argument(
+        "--redfish-discover-cache-ttl",
+        help="How long should the cache be stored in seconds (default: 86400) ",
```
```suggestion
        help="How long should the redfish discovery info stored in cache, in seconds (default: 86400) ",
```

dashmage replied:
Done!

sudeephb commented on prometheus_hardware_exporter/collectors/redfish.py:
```diff
@@ -38,28 +48,55 @@ def get_sensor_data(self, host: str, username: str, password: str) -> Dict[str,
+    def _get_sensor_data(
+        self, login_info: dict, timeout: int, max_retry: int
+    ) -> Optional[List[Any]]:
         """Return sensor if sensor exists else None."""
```
Can you also describe the parameters(especially login_info) in the docstring? Since it's now a dict, it's not very intuitive.

dashmage replied:
Good suggestion, adding for both the sensor methods.

sudeephb reviewed (commented):
Few nitpick comments. Otherwise LGTM.

rgildein commented on prometheus_hardware_exporter/collectors/redfish.py:
```diff
@@ -1,24 +1,41 @@
+        self.discover = get_discover(discover_cache_ttl)
 
-    def get_sensor_data(self, host: str, username: str, password: str) -> Dict[str, List]:
+    def get_sensor_data(self, login_info: dict, timeout: int, max_retry: int) -> Dict[str, List]:
```
not-blocking: I think it would be more readable to leave `host`, `username` and `password ` as arguments, instead of hiding it to one single argument.

dashmage replied:
The reason behind this was because there was a pylint error with too many arguments for the method (6 including `self`, it should be below 5). So I thought I'd collect all the redfish login related vars into one dict and pass that to the method.

Should I disable the pylint error instead and pass the vars separately?

jneo8 replied:
I am agree with Robert, `login_info` doesn't make things clear. We can try to ignore pylint here.

dashmage replied:
Since both of you are suggesting this, I'll revert my change, separate all the args again and add a "pylint disable" comment. IMO I think it's best to keep the number of arguments low to avoid too much complexity in the function and to bundle related arguments in a data structure so that it's easier to understand, maintain and test overall.

@rgildein - IMO it's good to keep the default number of max arguments (6) and not ignore this error by default from the template. Ideally 3-4 args would be good for a function and if it needs more, it's probably a sign that the function is doing too much work and we can split into another function. For this specific case, I can add the comment to disable to disable the pylint error.

rgildein replied:
> IMO I think it's best to keep the number of arguments low to avoid too much complexity in the function and to bundle related arguments in a data structure so that it's easier to understand, maintain and test overall.

If you wrap 3 arguments into single one and than unwrap it and use all those arguments + other in single call, it's not cleaner, easier to tests and deferentially not providing lower complexity.

```python
def foo(a: Dict[str, Any], b: int, c: int) -> None:
    a1 = a["a1"]
    a2 = a["a2"]
    a3 = a["a3"]
    foo2(a1, a2, a3, b, c)
```

I'm not sure that 6 is always enough, since we are using a lot 3-th party staff and our functions sometimes needs too many arguments, because they are forwarding them to the external function. I think that 9 will make much more sense. I'm doing lot of reviews and everyday I'm saying usage of pylint disable for this.

Btw, never use force push after review, because now I could not check what did you changed from my last review. Thanks

dashmage replied:
I understand the situation in your example where encapsulating all the args into a single one might not always have lesser complexity when it's easier to pass them all along separately. 

But I think as a general practice if the function is taking that many number of arguments (>6), it would be a bit complex to make sure to pass all the right args to the function each time you call it. Even during testing, we need to send all these args. But at the end of the day, it all depends on the exact situation :)

> Btw, never use force push after review, because now I could not check what did you changed from my last review.

Ah, it's a force of habit my bad. I don't want to mess up the log with lots of useless comments but I also understand we can just squash them before merging. Will make sure to not force push from now :+1:

rgildein replied:
> I understand the situation in your example where encapsulating all the args into a single one might not always have lesser complexity when it's easier to pass them all along separately.

I will not have any issue if you do 
```python
def foo(foo_arguments) -> None:
    """...
    params:
        foo_arguments: arguments for foo2
    """
    foo2(**foo_arguments)
```
Or simply
```python
def foo(*args, **kwargs) -> None:
    ...
    foo2(*args, **kwargs)
```

But in your case, when you never used `login_info` anywhere and simply pass those along with `timeout` and `max_retry` to sub-function, same function. Like this you only creates potential issue that someone will define `login_info` wrongly and got simple KeyError instead of `TypeError: 'a' is an invalid keyword argument for redfish.redfish_client`.

When you are using good IDE, it will tells you that you are missing e.g. `host` or that you provide wrong type for `host`, but when you do `login_info` it will tells you nothing.

dashmage replied:
After the latest changes, all the values are being sent to `RedfishHelper` as an instance of the `Config` object containing the login args (host, username, password) and the ttl value for the cache. This eliminates the need to send it all separately and also makes sure there is a tight coupling between the related data elements. It's also easier to instantiate the class now and prevents declaring an arbitrary `login_info` which isn't being used anywhere else.

rgildein commented on prometheus_hardware_exporter/collectors/redfish.py:
```diff
@@ -38,28 +56,61 @@ def get_sensor_data(self, host: str, username: str, password: str) -> Dict[str,
-    def discover(self) -> bool:
-        """Return true if redfish is been discovered."""
+
+def get_discover(ttl: int) -> Callable:
+    """Return the cached discover function.
+
+    Passes the ttl value to the cache decorator at runtime.
+    """
+
+    @ttl_cache(ttl=ttl)
+    def _discover() -> bool:
+        """Return true if redfish services have been discovered."""
+        logger.info("Discovering redfish services...")
         services = redfish.discover_ssdp()
         if len(services) == 0:
+            logger.info("No redfish services discovered")
             return False
+        logger.info("Discovered redfish services: %s", services)
         return True
+
+    return _discover
```
Why we moved this out of the class? Why not:
```python
def discover(self) -> bool:
    @ttl_cache(ttl=self.ttl)
    def _discover() -> bool:
        ...

    return _discover()
```

I do not understand why we are always creating class object, which ware basically storing no values. before you add `__init__` there was no class variable, no variable defined in __init__ and no properties.

jneo8 replied:
The example somehow make the `@ttl_cache` not working. The purpose here is just let the decorator can get the config value as the ttl settings.

rgildein replied:
I think we can easily do:
```python
@ttl_cache(ttl=TTF)
def discover(self) -> bool:
...
```
or completely remove this class object and define just simple functions outside of it.

My issue is that we define class and them moved function, which is 100% related with the class, outside of class object only to override bad design of cachetools and our. I do not think that this is blocking, but it's just bad design.

dashmage replied:
Thanks for the comment @rgildein !

The main aim here is to configure the ttl value for the cache at runtime which is coming from the the CLI args defined in `__main__.py`. I initially defined a class attribute which could then be modified at runtime to change the ttl value. But that didn't work since the decorator caches the initial value of the ttl class attribute when the class is initialized and doesn't change even when the attribute it modified later.

After some help from @jneo8 I defined a sort of wrapper function that returns the cached discover method with the ttl value being passed at runtime. 

I don't know why I didn't try this out earlier, but having the method inside the class itself seems to work while testing on a machine and it passes the unit tests as well. So we don't have to have the bad design after all :)

```python
class RedfishHelper:
    """Helper function for redfish."""

    def __init__(self, discover_cache_ttl: int) -> None:
        """Initialize disover method with TTL value."""
        self.discover = self.get_discover(discover_cache_ttl)

    (...)

    def get_discover(self, ttl: int) -> Callable:
        """Return the cached discover function.

        Passes the ttl value to the cache decorator at runtime.
        """

        @ttl_cache(ttl=ttl)
        def _discover() -> bool:
            """Return true if redfish services have been discovered."""
            logger.info("Discovering redfish services...")
            services = redfish.discover_ssdp()
            if len(services) == 0:
                logger.info("No redfish services discovered")
                return False
            logger.info("Discovered redfish services: %s", services)
            return True

        return _discover

rgildein replied:
Like this we are doing wrapper for wrapper, which is actually also wrapper. That is why I was trying to avoid using `get_discover`. But if there is no other solution, it's ok.

dashmage replied:
Yeah I really couldn't think of a more elegant solution to this problem either :disappointed:

jneo8 reviewed (approved):
LGTM

rgildein commented on prometheus_hardware_exporter/collectors/redfish.py:
```diff
@@ -1,24 +1,49 @@
 logger = getLogger(__name__)
 
 
+# pylint: disable=R0903
+# pylint: disable=R0913
```
not-blocking: I think you can use `R0903, R0913`

esunar replied:
can you use extended version of those so we can understand without checking what it means
`# pylint: disable=too-few-public-methods`

dashmage replied:
It also looks like I can omit `RO903` for too-few-public-methods since the `get_discover` method is added in the class now.

rgildein reviewed (approved):
LGTM

esunar commented on prometheus_hardware_exporter/collectors/redfish.py:
```diff
@@ -1,24 +1,49 @@
-    def get_sensor_data(self, host: str, username: str, password: str) -> Dict[str, List]:
+    def __init__(self, discover_cache_ttl: int) -> None:
+        """Initialize disover method with TTL value."""
+        self.discover = self.get_discover(discover_cache_ttl)
```
Why don't we initialize the helper with host: str, username: str, password: str, timeout: int, max_retry: int

esunar replied:
Or just with Config

dashmage replied:
That's a really nice idea. We wouldn't need to pass these login parameters each time to the get_sensor methods and while adding new methods for additional redfish metrics, their arg list would be simpler too. Will implement and push out this change :100:

esunar commented on prometheus_hardware_exporter/collectors/redfish.py:
```diff
@@ -38,28 +63,62 @@ def get_sensor_data(self, host: str, username: str, password: str) -> Dict[str,
 
-    def _get_sensor_data(self, host: str, username: str, password: str) -> Optional[List[Any]]:
-        """Return sensor if sensor exists else None."""
+    def _get_sensor_data(
```
two methods with same name?

rgildein replied:
Private and public?

dashmage replied:
Yep, `_get_sensor_data` is internal to the class while `get_sensor_data` actually returns data to the outside.

esunar reviewed (changes requested):
I've some inline comments and questions. Can you check please

dashmage commented:
Implemented very helpful changes suggested by @esunar . Also fixed the unit tests and added another one for the new method being implemented to map the sensor data.

rgildein reviewed (approved):
LGTM
I like idea with passing config to __init__ and now even class make much more sense :+1:. Good job.
